### DISCUSSION TITLE
Preferred endianness

### DISCUSSION THREAD

**Discussion by hackaugusto** [0xMiden/miden-vm#656] [General]
Some simple questions about endianness on the VM.

## Bytes enumeration

Given the word `W` with bytes `0x04 0x03 0x02 0x01`, in big-endian format (most significant byte is `0x04`). How should the bytes be named?

byte | 0x04 | 0x03 | 0x02 | 0x01 | description |
-- | ---- | ---- | ---- | ---- | -- |
1 | w[0] | w[1] | w[0] | w[3] | index is byte position from right-to-left (MSB is 0) |
**2** | **w[3]** | **w[2]** | **w[1]** | **w[0]** | **index is byte position from left-to-right (LSB is 0)  (*)** |

## Endianness on the stack

Would it be little endian or big endian?

stack | s[0] | s[1] | s[2] | s[3] | description |
-- | ---- | ---- | ---- | ---- | -- |
LE | 0x01 | 0x02 | 0x03 | 0x04 | LSB first |
**BE** | **0x04** | **0x03** | **0x02** | **0x01** | **MSB first (*)** |

`*` marks the preferred denotation

**Comment by bobbinth** [comment]
Great topic!

One note on terminology: we use "word" to denote an array of 4 field elements. So, words consist of field elements and not bytes. Field elements can be split up into bytes (8 bytes per element).

Regarding endianness: we typically use little-ending order. So, a word is `(a0, a1, a2, a3)` where `a0` is the least significant element of the word.

When a word is placed on the stack, it is arranged in "stack order" which is the reverse of the normal order. So, it appears on the stack as `[a3, a2, a1, a0, ...]` where `a3` is at the top of the stack. The reasoning for this is that if we were to read the word element-by-element and push each element onto the stack we'd first push `a0` onto the stack, then `a1` onto the stack etc.

Bytes within an element are assumed to be arranged in the little endian order as well. So, for example, when we use `u32split` operation (`[a, ...] -> [a1, a0, ...]`), the top 32 bits are put at the top of the stack, and lower 32 bits are put at the bottom of the stack.

We treat `u64` values in a similar manner. A `u64` value consisting of two 32-bit limbs `(a0, a1)` where `a0` is the least significant limb, would be placed on the stack as `[a1, a0, ...]`.
